File: python/problems/leetcode/0042_trapping_rain_water/solution.py
"""LeetCode 42: Trapping Rain Water."""

from __future__ import annotations


class Solution:
    def trap(self, height: list[int]) -> int:
        if not height:
            return 0
        l, r = 0, len(height) - 1
        lm = rm = 0
        ans = 0
        while l < r:
            if height[l] < height[r]:
                if height[l] >= lm:
                    lm = height[l]
                else:
                    ans += lm - height[l]
                l += 1
            else:
                if height[r] >= rm:
                    rm = height[r]
                else:
                    ans += rm - height[r]
                r -= 1
        return ans


if __name__ == "__main__":
    assert Solution().trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]) == 6
    assert Solution().trap([4, 2, 0, 3, 2, 5]) == 9
    print("0042 OK")

File: cpp/problems/leetcode/0042_trapping_rain_water/solution.cpp
// LeetCode 42
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        int n = static_cast<int>(height.size());
        if (n == 0) return 0;
        int l = 0, r = n - 1, lm = 0, rm = 0, ans = 0;
        while (l < r) {
            if (height[l] < height[r]) {
                if (height[l] >= lm)
                    lm = height[l];
                else
                    ans += lm - height[l];
                ++l;
            } else {
                if (height[r] >= rm)
                    rm = height[r];
                else
                    ans += rm - height[r];
                --r;
            }
        }
        return ans;
    }
};

int main() {
    vector<int> a{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    assert(Solution().trap(a) == 6);
    vector<int> b{4, 2, 0, 3, 2, 5};
    assert(Solution().trap(b) == 9);
    cout << "0042 OK" << endl;
    return 0;
}
